### PR TITLE
feat: support created_at filter for the GET /users endpoint

### LINKED ISSUE #12747: support `created_at` filter for the `GET /users` endpoint
we support these filters currently: https://coder.com/docs/v2/latest/admin/users#user-filtering

a major customer would like to filter by `created_at`

### PR DESCRIPTION
**Pull request by joobisb**
Closes https://github.com/coder/coder/issues/12747

We support these filters currently: https://coder.com/docs/v2/latest/admin/users#user-filtering, adding `created_at` filter as well.

### CODE REVIEW

**Comment by joobisb** [comment]
@ericpaulsen @matifali could you please take a look at this?

**Comment by joobisb** [comment]
> Thanks for the PR!
> 
> Whilst this technically addresses the linked issue, I'm not sure how useful it is if it doesn't support date ranges. I think it'd make more sense to implement it in terms of `created_at_before` and `created_at_after`, as we do for `last_seen_*`. You could obviously then use a combination of the two to retrieve users created on a specific day.
> 
> Although not mentioned in the issue, we'll also want to support this filter on the frontend (a separate PR for that is fine), on the `deployment/users` page.

@ethanndickson I've updated the PR to support date ranges. Also I don't think we need to make any changes in the frontend as the params are directly passed on to the backend and is working fine as shown below.

Please do have a look at the PR.

image

File: docs/admin/users/index.md [hunk 1]
```diff
@@ -195,6 +197,8 @@ The following filters are supported:
 - `role` - Represents the role of the user. You can refer to the
   [TemplateRole documentation](https://pkg.go.dev/github.com/coder/coder/v2/codersdk#TemplateRole)
   for a list of supported user roles.
-- `last_seen_before` and `last_seen_after` - The last time a used has used the
+- `last_seen_before` and `last_seen_after` - The last time a user has used the
   platform (e.g. logging in, any API requests, connecting to workspaces). Uses
   the RFC3339Nano format.
+- `created_at_before` and `created_at_after` - The time a user was created. Uses
```
Comment (ethanndickson) [inline docs/admin/users/index.md:203]: I don't think `created_at_{after|before}` reads very well. I didn't really think about the phrasing when I wrote it earlier. Let's just go with `created_before` and `created_after`.

**Review by ethanndickson — COMMENTED** [review]
LGTM, just one comment on naming, sorry for misleading you there!

File: docs/admin/users/index.md (hunk 1 above)
Comment (joobisb) [reply docs/admin/users/index.md]: @ethanndickson 

updated the PR!

File: coderd/users_test.go [hunk 2]
```diff
@@ -1655,6 +1655,37 @@ func TestUsersFilter(t *testing.T) {
...
+			FilterF: func(_ codersdk.UsersRequest, u codersdk.User) bool {
+				start := time.Date(2023, 1, 1, 0, 0, 0, 0, time.UTC)
+				return u.CreatedAt.After(start)
+			},
+		},
+		{
+			Name: "CreatedAtRange",
+			Filter: codersdk.UsersRequest{
+				SearchQuery: `created_after:"2023-01-01T00:00:00Z" created_before:"2023-01-31T23:59:59Z"`,
+			},
```
Comment (ethanndickson) [inline coderd/users_test.go:1682]: All the users created in the setup of this test have the same creation date. We should add some users created within the range, and before the `CreatedBefore` testcase timestamp. Right now those testcases are just comparing empty slices.

Comment (joobisb) [reply coderd/users_test.go:1719]: @ethanndickson 

I've updated the PR with the tests, please do have a look

File: coderd/users_test.go [hunk 3]
```diff
@@ -1521,10 +1560,8 @@ func TestUsersFilter(t *testing.T) {
 		FilterF func(f codersdk.UsersRequest, user codersdk.User) bool
 	}{
 		{
-			Name: "All",
-			Filter: codersdk.UsersRequest{
-				Status: codersdk.UserStatusSuspended + "," + codersdk.UserStatusActive,
-			},
+			Name:   "All",
+			Filter: codersdk.UsersRequest{},
```
Comment (joobisb) [inline coderd/users_test.go:1564]: updated this since "All" filter should include all the users

Comment (ethanndickson) [reply coderd/users_test.go:1564]: Given this is done twice, I assume this was intentional. Just a sanity check that all users are either suspended or active. I think we should just keep the original.

File: coderd/users_test.go [hunk 4]
```diff
@@ -1602,7 +1639,7 @@ func TestUsersFilter(t *testing.T) {
 				Status: codersdk.UserStatusSuspended + "," + codersdk.UserStatusActive,
 			},
 			FilterF: func(_ codersdk.UsersRequest, u codersdk.User) bool {
-				return true
+				return u.Status == codersdk.UserStatusSuspended || u.Status == codersdk.UserStatusActive
```
[the remark this answers is not in the document]
Comment (ethanndickson) [reply coderd/users_test.go:1642]: Same here, let's keep the original as it was almost certainly purposeful.

File: coderd/users_test.go [hunk 5]
```diff
@@ -1513,6 +1513,45 @@ func TestUsersFilter(t *testing.T) {
 		users = append(users, user)
 	}
 
+	// Add users with different creation dates for testing date filters
+	for i := 0; i < 3; i++ {
+		// nolint:gocritic
```
Comment (ethanndickson) [inline coderd/users_test.go:1518]: Each of these `nolint` should mention that the system context is necessary here. (Also to make it clear which specific rule this is `nolint`ing)

**Review by ethanndickson — COMMENTED** [review]
Few comments then I'm happy to approve!

**Comment by joobisb** [comment]
> Few comments then I'm happy to approve!

@ethanndickson addressed the comments!

File: coderd/users_test.go [hunk 6]
```diff
@@ -2273,6 +2346,28 @@ func onlyUsernames[U codersdk.User | database.User](users []U) []string {
 	return out
 }
 
+// dbUserToSDKUser converts database.User to codersdk.User
+func dbUserToSDKUser(dbUser database.User) codersdk.User {
```
Comment (ethanndickson) [inline coderd/users_test.go:2350]: We fortunately already have this function, `db2sdk.User`. It takes a `database.User` and a slice of organization IDs. In this case that slice should just be `[]uuid.UUID{first.OrganizationID})`

**Review by ethanndickson — COMMENTED** [review]
Last comment, thanks for your patience!

File: coderd/users_test.go (hunk 6 above)
Comment (joobisb) [reply coderd/users_test.go:2350]: thanks for pointing this out!

**Comment by joobisb** [comment]
> Last comment, thanks for your patience!

@ethanndickson Done!

File: coderd/users_test.go [hunk 7]
```diff
@@ -1675,6 +1749,8 @@ func TestUsersFilter(t *testing.T) {
 					exp = append(exp, made)
 				}
 			}
+			fmt.Printf("expexp: %+v\n", exp)
+			fmt.Printf("matched.Usersmatched.Users: %+v\n", matched.Users)
```
Comment (ethanndickson) [inline coderd/users_test.go:1753]:
```suggestion
```

Comment (joobisb) [reply coderd/users_test.go]: my bad, fixed it!

**Comment by ethanndickson** [comment]
Looks like the test is failing on Postgres.

The expected list of users have timestamps with time zones, while the actual list of users retrieved via the API has timestamps parsed from strings, which don’t have timezones. I’m not sure of the best way to fix that..

**Comment by joobisb** [comment]
> Looks like the test is failing on Postgres.
> 
> The expected list of users have timestamps with time zones, while the actual list of users retrieved via the API has timestamps parsed from strings, which don’t have timezones. I’m not sure of the best way to fix that..

@ethanndickson put together a little hack to fix this. Instead of using the db object, call an API to update the status and used that user object to be added to the list of expected users. Please do have a look.

**Comment by ethanndickson** [comment]
Yep, definitely unrelated. Will fix.

**Comment by ethanndickson** [comment]
Should be fixed after a merge from main!

File: coderd/users_test.go [hunk 8]
```diff
@@ -1677,6 +1777,17 @@ func TestUsersFilter(t *testing.T) {
 					exp = append(exp, made)
 				}
 			}
+
+			// TODO: This can be removed with dbmem
+			if !dbtestutil.WillUsePostgres() {
```
Comment (ethanndickson) [inline coderd/users_test.go:1782]: The plan is to eventually get rid of `dbmem` entirely, so I'm not too worried about cluttering up this test. This is to handle a disparity between organization-less members in Postgres and dbmem. (FWIW it's not possible to have organization-less members on a real Coder deployment.)

**Review by ethanndickson — APPROVED** [review]
Thanks for your contribution and your patience through the review cycle! Greatly appreciated!

**Pull request merged**